File: server-deno/cloudflare.ts
import { Redis } from "https://deno.land/x/redis@v0.13.1/mod.ts";

export interface CheckResult {
    time: number;
    responseTime: number;

    serverHeader: boolean;   // server-deno: cloudflare
    // See https://support.cloudflare.com/hc/en-us/articles/200170156-Understanding-the-Cloudflare-Cookies
    duidCookie: boolean;                // __cfduid
    loadBalancerCookie: boolean;        // __cflb
    botManagementCookie: boolean;       // __cf_bm
    offlineBrowsingInfoCookie: boolean; // cf_ob_info
    useOfflineBrowsingCookie: boolean;  // cf_use_ob
}

export async function getLastCheck(redis: Redis, hostname: string): Promise<CheckResult | null> {
    const [
        timeString,
        responseTimeString,
        serverHeaderString,
        duidCookieString,
        loadBalancerCookieString,
        botManagementCookieString,
        offlineBrowsingInfoCookieString,
        useOfflineBrowsingCookieString,
    ] = await redis.hmget(
        `previous-check:hostname#${hostname}`,
        "time",
        "response-time",
        "server-header",
        "duid-cookie",
        "load-balancer-cookie",
        "bot-management-cookie",
        "offline-browsing-info-cookie",
        "use-offline-browsing-cookie",
    );
    if(timeString === undefined) {
        return null;
    }

    return {
        time: Number(timeString),
        responseTime: Number(responseTimeString),
        serverHeader: serverHeaderString === "1",
        duidCookie: duidCookieString === "1",
        loadBalancerCookie: loadBalancerCookieString === "1",
        botManagementCookie: botManagementCookieString === "1",
        offlineBrowsingInfoCookie: offlineBrowsingInfoCookieString === "1",
        useOfflineBrowsingCookie: useOfflineBrowsingCookieString === "1",
    };
}

export async function updateLastCheck(redis: Redis, hostname: string, check: CheckResult) {
    const boolToBitString = (bool: boolean) => bool ? "1" : "0";

    await redis.hset(
        `previous-check:hostname#${hostname}`,
        ["time", `${check.time}`],
        ["response-time", `${check.responseTime}`],
        ["server-header", boolToBitString(check.serverHeader)],
        ["duid-cookie", boolToBitString(check.duidCookie)],
        ["load-balancer-cookie", boolToBitString(check.loadBalancerCookie)],
        ["bot-management-cookie", boolToBitString(check.botManagementCookie)],
        ["offline-browsing-info-cookie", boolToBitString(check.offlineBrowsingInfoCookie)],
        ["use-offline-browsing-cookie", boolToBitString(check.useOfflineBrowsingCookie)],
    );
}

// Wrapper for fetch with a timeout
function timeoutFetch(url: string, time: number): ReturnType<typeof fetch> {
    return new Promise<Response>(async(resolve, reject) => {
        const timeout = setTimeout(() => {
            reject(new Error(`Timeout after ${time}ms`));
        }, time);

        const fetchPromise = fetch(url);
        try {
            const response = await fetchPromise;
            // The resolve gets called if there is a response after the timeout, but it gets ignored by the caller
            resolve(response);
        } catch(err) {
            reject(err);
        } finally {
            clearInterval(timeout);
        }
    });
}

export async function makeRequestAndCheck(url: string): Promise<CheckResult> {
    const startTime = Date.now();
    const response = await timeoutFetch(url, 10_000);
    const endTime = Date.now();
    const indicators: CheckResult = {
        time: Date.now(),
        responseTime: endTime - startTime,
        serverHeader: false,
        duidCookie: false,
        loadBalancerCookie: false,
        botManagementCookie: false,
        offlineBrowsingInfoCookie: false,
        useOfflineBrowsingCookie: false,
    };

    if(response.headers.get('server-deno') === 'cloudflare') {
        indicators.serverHeader = true;
    }
    const setCookieHeader = response.headers.get('Set-Cookie');
    if(setCookieHeader !== null) {
        indicators.duidCookie = setCookieHeader.includes('__cfduid=');
        indicators.loadBalancerCookie = setCookieHeader.includes('__cflb=');
        indicators.botManagementCookie = setCookieHeader.includes('__cf_bm=');
        indicators.offlineBrowsingInfoCookie = setCookieHeader.includes('cf_ob_info=');
        indicators.useOfflineBrowsingCookie = setCookieHeader.includes('cf_use_ob=');
    }

    return indicators;
}

File: server-golang/cloudflare.go
package main

import (
	"context"
	"fmt"
	"github.com/go-redis/redis/v8"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type checkResult struct {
	Time int64						`json:"time"`
	ResponseTime int64				`json:"responseTime"`

	ServerHeader bool               `json:"serverHeader"`
	// See https://support.cloudflare.com/hc/en-us/articles/200170156-Understanding-the-Cloudflare-Cookies
	DuidCookie bool                 `json:"duidCookie"`                 // __cfduid
	LoadBalancerCookie bool         `json:"loadBalancerCookie"`         // __cflb
	BotManagementCookie bool        `json:"botManagementCookie"`        // __cf_bm
	OfflineBrowsingInfoCookie bool  `json:"offlineBrowsingInfoCookie"`  // cf_ob_info
	UseOfflineBrowsingCookie bool   `json:"useOfflineBrowsingCookie"`   // cf_use_ob
}

func getLastCheck(ctx context.Context, redis *redis.Client, hostname string) (*checkResult, error) {
	res, err := redis.HMGet(
		ctx,
		fmt.Sprintf("previous-check:hostname#%s", hostname),
		"time",
		"response-time",
		"server-header",
		"duid-cookie",
		"load-balancer-cookie",
		"bot-management-cookie",
		"offline-browsing-info-cookie",
		"use-offline-browsing-cookie",
	).Result()
	if err != nil { // redis returns error
		return nil, err
	}
	if res[0] == nil { // hostname has not been checked before so results are empty
		return nil, nil
	}

	t, err := strconv.ParseInt(res[0].(string), 10, 64)
	if err != nil {
		return nil, err
	}
	respt, err := strconv.ParseInt(res[1].(string), 10, 64)
	if err != nil {
		return nil, err
	}

	return &checkResult{
		Time:                      t,
		ResponseTime:			   respt,
		ServerHeader:              res[2].(string) == "1",
		DuidCookie:                res[3].(string) == "1",
		LoadBalancerCookie:        res[4].(string) == "1",
		BotManagementCookie:       res[5].(string) == "1",
		OfflineBrowsingInfoCookie: res[6].(string) == "1",
		UseOfflineBrowsingCookie:  res[7].(string) == "1",
	}, nil
}

func updateLastCheck(ctx context.Context, redis *redis.Client, hostname string, check *checkResult) error {
	_, err := redis.HMSet(
		ctx,
		fmt.Sprintf("previous-check:hostname#%s", hostname),
		"time",
		check.Time,
		"response-time",
		check.ResponseTime,
		"server-header",
		check.ServerHeader,
		"duid-cookie",
		check.DuidCookie,
		"load-balancer-cookie",
		check.LoadBalancerCookie,
		"bot-management-cookie",
		check.BotManagementCookie,
		"offline-browsing-info-cookie",
		check.OfflineBrowsingInfoCookie,
		"use-offline-browsing-cookie",
		check.UseOfflineBrowsingCookie,
	).Result()
	return err
}

func check(res *http.Response) *checkResult {
	indicators := &checkResult{
		Time: time.Now().Unix(),
		ResponseTime: 0,

		ServerHeader:              false,
		DuidCookie:                false,
		LoadBalancerCookie:        false,
		BotManagementCookie:       false,
		OfflineBrowsingInfoCookie: false,
		UseOfflineBrowsingCookie:  false,
	}

	if res.Header.Get("server") == "cloudflare" {
		indicators.ServerHeader = true
	}
	setCookieHeader := res.Header.Get("Set-Cookie")
	if setCookieHeader != "" {
		indicators.DuidCookie = strings.Contains(setCookieHeader, "__cfduid=")
		indicators.LoadBalancerCookie = strings.Contains(setCookieHeader, "__cflb=")
		indicators.BotManagementCookie = strings.Contains(setCookieHeader, "__cf_bm=")
		indicators.OfflineBrowsingInfoCookie = strings.Contains(setCookieHeader, "cf_ob_info=")
		indicators.UseOfflineBrowsingCookie = strings.Contains(setCookieHeader, "cf_use_ob=")
	}

	return indicators
}

func makeRequestAndCheck(url string) (*checkResult, error) {
	req, err := http.NewRequest("GET", url, nil)
	// Couldn't create request - maybe invalid url
	if err != nil {
		return nil, err
	}

	c := &http.Client{
		Timeout: 5 * time.Second,
	}

	start := time.Now()
	res, err := c.Do(req)
	// Error during the request - could be timeout
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	checkRes := check(res)
	checkRes.ResponseTime = time.Since(start).Microseconds()

	return checkRes, nil
}
